src/find/matchers/mod.rs

```diff
@@ -0,0 +1,732 @@
+mod logical_matchers;
+mod name;
+mod printer;
+mod prune;
+mod size;
+mod time;
+mod type_matcher;
+
+use regex::Regex;
+use std::error::Error;
+use std::time::SystemTime;
+use walkdir::DirEntry;
+
+use find::{Config, Dependencies};
+
+
+
+/// Struct holding references to outputs and any inputs that can't be derived
+/// from the file/directory info.
+pub struct MatcherIO<'a> {
+    should_skip_dir: bool,
+    deps: &'a Dependencies<'a>,
+}
+
+impl<'a> MatcherIO<'a> {
+    pub fn new(deps: &'a Dependencies<'a>) -> MatcherIO<'a> {
+        MatcherIO {
+            deps: deps,
+            should_skip_dir: false,
+        }
+    }
+
+    pub fn mark_current_dir_to_be_skipped(&mut self) {
+        self.should_skip_dir = true;
+    }
+
+    pub fn should_skip_current_dir(&self) -> bool {
+        self.should_skip_dir
+    }
+
+    pub fn now(&self) -> SystemTime {
+        self.deps.now()
+    }
+}
+
+/// A basic interface that can be used to determine whether a directory entry
+/// is what's being searched for. To a first order approximation, find consists
+/// of building a chain of Matcher objets, and then walking a directory tree,
+/// passing each entry to the chain of Matchers.
+pub trait Matcher {
+    /// Returns whether the given file matches the object's predicate.
+    fn matches(&self, file_info: &DirEntry, matcher_io: &mut MatcherIO) -> bool;
+
+    /// Returns whether the matcher has any side-effects. Iff no such matcher
+    /// exists in the chain, then the filename will be printed to stdout. While
+    /// this is a compile-time fact for most matchers, it's run-time for matchers
+    /// that contain a collection of sub-Matchers.
+    fn has_side_effects(&self) -> bool;
+}
+
+pub enum ComparableValue {
+    MoreThan(u64),
+    EqualTo(u64),
+    LessThan(u64),
+}
+
+impl ComparableValue {
+    fn matches(&self, value: u64) -> bool {
+        match *self {
+            ComparableValue::MoreThan(limit) => value > limit,
+            ComparableValue::EqualTo(limit) => value == limit, 
+            ComparableValue::LessThan(limit) => value < limit, 
+        }
+    }
+
+    /// same as matches, but takes a signed value
+    fn imatches(&self, value: i64) -> bool {
+        match *self {
+            ComparableValue::MoreThan(limit) => value >= 0 && (value as u64) > limit,
+            ComparableValue::EqualTo(limit) => value >= 0 && (value as u64) == limit, 
+            ComparableValue::LessThan(limit) => value < 0 || (value as u64) < limit, 
+        }
+    }
+}
+
+
+/// Builds a single AndMatcher containing the Matcher objects corresponding
+/// to the passed in predicate arguments.
+pub fn build_top_level_matcher(args: &[&str],
+                               config: &mut Config)
+                               -> Result<Box<Matcher>, Box<Error>> {
+    let (_, top_level_matcher) = try!(build_matcher_tree(args, config, 0, false));
```

Comment:
My personal feeling is that ? is now standard and should be preferred. This is a point of style, though.